### ISSUE TITLE
Cross-host connects not showing in UI when connection traverses docker port mapping

### ISSUE THREAD

**Issue by tomwilkie** [weaveworks/scope#227]
![conn](https://cloud.githubusercontent.com/assets/444037/8158888/5e0fb7aa-1359-11e5-9399-18658c96e1ac.png)

**Comment by tomwilkie** [comment]
Contents of /api/report

https://gist.github.com/tomwilkie/bfb32be87d63adc6cc3f

**Comment by tomwilkie** [comment]
```
oot     27318  0.0  0.0    768     4 ?        Ss   11:15   0:00  \_ /sbin/runsvdir /etc/service
root     27335  0.0  0.0    744     4 ?        S    11:15   0:00      \_ runsv probe
root     27338  5.5  0.3  17176 14720 ?        Sl   11:15   0:32      |   \_ /home/weave/probe
root     27336  0.0  0.0    744     4 ?        S    11:15   0:00      \_ runsv app
root     27337  0.2  0.3  16792 13824 ?        Sl   11:15   0:01          \_ /home/weave/app 10.240.240.51
vagrant@twilkie-test-1:~/src/github.com/weaveworks/scope$ sudo su
root@twilkie-test-1:/home/vagrant/src/github.com/weaveworks/scope# cat /proc/27338/
attr/            cmdline          environ          io               mem              ns/              pagemap          schedstat        stat             timers
autogroup        comm             exe              limits           mountinfo        numa_maps        personality      sessionid        statm            uid_map
auxv             coredump_filter  fd/              loginuid         mounts           oom_adj          projid_map       setgroups        status           wchan
cgroup           cpuset           fdinfo/          map_files/       mountstats       oom_score        root/            smaps            syscall          
clear_refs       cwd/             gid_map          maps             net/             oom_score_adj    sched            stack            task/            
root@twilkie-test-1:/home/vagrant/src/github.com/weaveworks/scope# cat /proc/27338/net/tcp
  sl  local_address rem_address   st tx_queue rx_queue tr tm->when retrnsmt   uid  timeout inode                                                     
   0: 00000000:0016 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 10811 1 ffff880129778e00 100 0 0 10 0                     
   1: 0100007F:8F96 0100007F:0FBE 01 00000000:00000000 00:00000000 00000000     0        0 88240 1 ffff8801293f5c00 20 4 10 10 -1                    
   2: E251F00A:9261 33F0F00A:0FBE 01 00000000:00000000 00:00000000 00000000     0        0 88941 1 ffff8800bae86a00 20 4 30 10 -1                    
   3: E251F00A:D047 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 89686 1 ffff8801293f4700 20 4 0 10 -1                     
   4: E251F00A:D36B FEA9FEA9:0050 01 00000000:00000000 00:00000000 00000000     0        0 93276 1 ffff880121520700 22 0 0 10 -1                     
   5: E251F00A:CF8B E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 88782 1 ffff8801293f4e00 20 4 0 10 -1                     
   6: E251F00A:CFAC E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 88947 1 ffff8800bae84000 20 4 0 10 -1                     
   7: E251F00A:D186 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 91137 1 ffff880121522300 20 4 0 10 -1                     
   8: E251F00A:D36A FEA9FEA9:0050 06 00000000:00000000 03:000004E6 00000000     0        0 0 3 ffff8800b9ddd500                                      
   9: E251F00A:D259 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 92104 1 ffff880121521500 20 4 0 10 -1                     
  10: E251F00A:CF6E E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 88246 1 ffff8801293f6300 20 4 0 10 -1                     
  11: E251F00A:D396 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 93805 1 ffff8800bae87800 20 4 0 10 -1                     
  12: E251F00A:0016 4C219D25:9E32 01 00000000:00000000 02:0005D377 00000000     0        0 15654 3 ffff880129778000 21 4 33 10 13                    
  13: E251F00A:CF7C E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 88712 1 ffff8801293f4000 20 4 0 10 -1                     
  14: E251F00A:D369 FEA9FEA9:0050 06 00000000:00000000 03:000004E5 00000000     0        0 0 3 ffff8800b9ddd400                                      
  15: E251F00A:D356 FEA9FEA9:0050 01 00000000:00000000 00:00000000 00000000     0        0 93186 1 ffff880121520000 20 0 0 10 -1                     
  16: E251F00A:D32B E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 93098 1 ffff880121522a00 20 4 0 10 -1                     
  17: E251F00A:9227 33F0F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 88247 1 ffff8801293f5500 20 4 28 10 -1                    
  18: E251F00A:D2C2 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 92592 1 ffff880121523100 20 4 0 10 -1                     
  19: E251F00A:D0B2 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 90172 1 ffff88012977aa00 20 4 0 10 -1                     
  20: E251F00A:CFE2 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 89198 1 ffff8800bae85c00 20 4 0 10 -1                     
  21: E251F00A:D11B E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 90647 1 ffff8800bae84700 20 4 0 10 -1                     
  22: E251F00A:CF75 E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 88680 1 ffff8801293f6a00 20 4 0 10 -1                     
  23: E251F00A:D1EF E251F00A:0FBE 08 00000000:00000000 00:00000000 00000000     0        0 91620 1 ffff880121523800 20 4 0 10 -1  
```

**Comment by tomwilkie** [comment]
E251F00A = 10.240.81.226
33F0F00A = 10.240.240.51

**Comment by tomwilkie** [comment]
This line:

```
 2: E251F00A:9261 33F0F00A:0FBE 01 00000000:00000000 00:00000000 00000000     0        0 88941 1 ffff8800bae86a00 20 4 30 10 -1                    
```

Translates as 10.240.81.226:37473 -> 10.240.240.51:4030

**Comment by tomwilkie** [comment]
From /api/report:

```
      "\u003etwilkie-test-1;10.240.81.226;37473": ["theinternet"],
```

shouldn't be theinternet.  Looks like we're over-squashing.

**Comment by tomwilkie** [comment]
/32?

```
  "Host": {
    "Adjacency": {},
    "EdgeMetadatas": {},
    "NodeMetadatas": {
      "twilkie-test-1;\u003chost\u003e": {
        "host_name": "twilkie-test-1",
        "load": "0.06 0.19 0.22",
        "local_networks": "127.0.0.1/8 10.240.81.226/32 172.17.42.1/16",
        "os": "linux",
        "ts": "2015-06-15T11:24:01.278056215Z"
      },
      "twilkie-test-2;\u003chost\u003e": {
        "host_name": "twilkie-test-2",
        "load": "0.00 0.03 0.05",
        "local_networks": "127.0.0.1/8 10.240.240.51/32 172.17.42.1/16",
        "os": "linux",
        "ts": "2015-06-15T11:24:02.63945769Z"
      }
    }
  }
```

```
vagrant@twilkie-test-2:~/src/github.com/weaveworks/scope$ ifconfig
docker0   Link encap:Ethernet  HWaddr 56:84:7a:fe:97:99  
          inet addr:172.17.42.1  Bcast:0.0.0.0  Mask:255.255.0.0
          UP BROADCAST RUNNING MULTICAST  MTU:1460  Metric:1
          RX packets:1418 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1800 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:639419 (639.4 KB)  TX bytes:1700473 (1.7 MB)

eth0      Link encap:Ethernet  HWaddr 42:01:0a:f0:f0:33  
          inet addr:10.240.240.51  Bcast:10.240.240.51  Mask:255.255.255.255
          UP BROADCAST RUNNING MULTICAST  MTU:1460  Metric:1
          RX packets:64581 errors:0 dropped:0 overruns:0 frame:0
          TX packets:28862 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:288899477 (288.8 MB)  TX bytes:14307900 (14.3 MB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:1740 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1740 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:8223682 (8.2 MB)  TX bytes:8223682 (8.2 MB)

```

**Comment by tomwilkie** [comment]
From hipchat: looks like we introduced a regression where we add a scope to non-local addresses.

Peter is putting together a fix.

**Comment by tomwilkie** [comment]
Still not working for me: http://104.155.2.227:4040/#!/state/{"topologyId":"containers","grouping":"none","selectedNodeId":null}

**Comment by tomwilkie** [comment]
So one reason its not working is that GCE has some sort of NAT in the way;  my laptop sees a different address to the probe on the GCE VM.

This still doesn't explain why GCE<->GCE isn't working.

**Comment by tomwilkie** [comment]
FFS looks like GCE does some NAT'ing between VMs:

```
2015/06/15 14:13:26 squashing >;10.240.240.51;47861 -> ;10.240.81.226;4030
2015/06/15 14:13:26 squashing >;10.240.240.51;57121 -> ;169.254.169.254;80
2015/06/15 14:13:26 squashing >;10.240.240.51;22 -> ;37.157.33.76;13222
2015/06/15 14:13:26 squashing >;10.240.240.51;4030 -> ;10.240.81.226;50088
2015/06/15 14:13:26 squashing >;10.240.240.51;4040 -> ;37.157.33.76;18440
2015/06/15 14:13:26 squashing >;10.240.240.51;4040 -> ;37.157.33.76;22082
2015/06/15 14:13:26 squashing >;10.240.240.51;4040 -> ;37.157.33.76;43281
2015/06/15 14:13:26 squashing >;10.240.240.51;57116 -> ;169.254.169.254;80
2015/06/15 14:13:26 squashing >;172.17.0.2;57476 -> ;10.240.81.226;6783
2015/06/15 14:13:26 squashing >;172.17.0.2 -> ;10.240.81.226
2015/06/15 14:13:26 squashing >;10.240.240.51 -> ;10.240.81.226
2015/06/15 14:13:26 squashing >;10.240.240.51 -> ;169.254.169.254
2015/06/15 14:13:26 squashing >;10.240.240.51 -> ;37.157.33.76
2015/06/15 14:13:26 connection initiated: 10.240.240.51:48269 (348ca20a07fc252f)
2015/06/15 14:13:26 connection initiated: 37.157.33.76:20683 (3de6933f47173e60)
2015/06/15 14:13:29 squashing >;10.240.81.226;22 -> ;37.157.33.76;40498
2015/06/15 14:13:29 squashing >;10.240.81.226;4040 -> ;37.157.33.76;28941
2015/06/15 14:13:29 squashing >;10.240.240.51;4040 -> ;37.157.33.76;18440
2015/06/15 14:13:29 squashing >;10.240.240.51;4040 -> ;37.157.33.76;22082
2015/06/15 14:13:29 squashing >;10.240.240.51;4040 -> ;37.157.33.76;43281
2015/06/15 14:13:29 squashing >;10.240.240.51;57121 -> ;169.254.169.254;80
2015/06/15 14:13:29 squashing >;10.240.240.51;57116 -> ;169.254.169.254;80
2015/06/15 14:13:29 squashing >;10.240.240.51;22 -> ;37.157.33.76;13222
2015/06/15 14:13:29 squashing >;10.240.81.226;4030 -> ;37.157.33.76;20683
2015/06/15 14:13:29 squashing >;10.240.81.226;41762 -> ;169.254.169.254;80
2015/06/15 14:13:29 squashing >;10.240.81.226;41754 -> ;169.254.169.254;80
2015/06/15 14:13:29 squashing >;10.240.81.226 -> ;169.254.169.254
2015/06/15 14:13:29 squashing >;10.240.81.226 -> ;37.157.33.76
2015/06/15 14:13:29 squashing >;10.240.240.51 -> ;169.254.169.254
2015/06/15 14:13:29 squashing >;10.240.240.51 -> ;37.157.33.76
2015/06/15 14:13:32 squashing >;10.240.240.51;57121 -> ;169.254.169.254;80
2015/06/15 14:13:32 squashing >;10.240.81.226;41762 -> ;169.254.169.254;80
2015/06/15 14:13:32 squashing >;10.240.81.226;41754 -> ;169.254.169.254;80
2015/06/15 14:13:32 squashing >;10.240.240.51;4040 -> ;37.157.33.76;18440
2015/06/15 14:13:32 squashing >;10.240.240.51;4040 -> ;37.157.33.76;22082
2015/06/15 14:13:32 squashing >;10.240.240.51;4040 -> ;37.157.33.76;43281
2015/06/15 14:13:32 squashing >;10.240.240.51;57116 -> ;169.254.169.254;80
2015/06/15 14:13:32 squashing >;10.240.81.226;22 -> ;37.157.33.76;40498
2015/06/15 14:13:32 squashing >;10.240.240.51;22 -> ;37.157.33.76;13222
2015/06/15 14:13:32 squashing >;10.240.81.226;4040 -> ;37.157.33.76;28941
2015/06/15 14:13:32 squashing >;10.240.81.226;4040 -> ;37.157.33.76;57299
2015/06/15 14:13:32 squashing >;10.240.81.226;4030 -> ;37.157.33.76;20683
2015/06/15 14:13:32 squashing >;10.240.81.226 -> ;169.254.169.254
2015/06/15 14:13:32 squashing >;10.240.81.226 -> ;37.157.33.76
2015/06/15 14:13:32 squashing >;10.240.240.51 -> ;169.254.169.254
2015/06/15 14:13:32 squashing >;10.240.240.51 -> ;37.157.33.76
```

**Comment by tomwilkie** [comment]
Last but not least, the weave <-> weave connections are not working.  I'm guessing this is all about dockers port remapping...

**Comment by peterbourgon** [comment]
This ticket exposed a few bugs, we fixed a couple for 0.3.0, the remaining work is to figure out a way to associate internal (within a container) ports with their external (host system) mappings. Created #245 for that and closing this one.
